### ISSUE TITLE
Behavior of first line indentation in paragraphs seems limiting

### ISSUE THREAD

**Issue by araruna** [typst/typst#311]
As I gathered [from the docs](https://typst.app/docs/reference/layout/par/) (as seen on the reference for the `first-line-indent` argument), the first paragraph on a page is special-cased to never have it's first line indented.

If I understand this correctly, that means not all paragraphs will have indentation applied, just like LaTeX does by default. If this is correct, then some styles will not be able to be represented with Typst. For example, in Brazilian Portuguese documents we are used to apply indentation to every paragraph, regardless of it's length or position in the text, and as such we use the LaTeX package `indentfirst` pervasively.

Did I understand it correctly?

**Comment by laurmaedje** [comment]
We'll add some configuration for this, so that both ways are possible.

**Comment by chris-morgan** [comment]
Possibly related, paragraphs starting with something other than unformatted text don’t get indented (e.g. `#super[x]`). All of this makes the feature feel temperamental.

**Comment by laurmaedje** [comment]
The bug with unformatted text not being indented should be fixed since the last release. Can you still reproduce it? More generally, I tried to fix the first line indent, but its not as simple as it seems. As currently _all_ text is layouted as a paragraph in Typst, removing the rule would indent _everything._ That includes headings, code blocks, captions, words in boxes, etc. Since paragraphs are not clearly demarked (as they would be in, say HTML), we first need some automatic way to distinguish these two kinds of texts (possibly through the context they appear in).

**Comment by zhonghes** [comment]
this problem is still exist
![image](https://user-images.githubusercontent.com/114350562/231048235-1b7637cc-36e8-434c-aa6b-d804850270bf.png)

**Comment by laurmaedje** [comment]
Yes, I know. As mentioned in my previous response, the overall issue is difficult to fix. The only thing that was fixed so far is that formatted text wasn't even indented when it wasn't in the first paragraph.

**Comment by DawnEver** [comment]
I found another para `hanging-indent`
if you want to indent first line in 1 em
```typst
set par(justify: false, leading: 1.5em, first-line-indent:2em )
```
you can do like this:
```typst
set par( hanging-indent:-1em )
```

**Comment by wychwitch** [comment]
@DawnEver oh wow, I never thought if using a negative hanging indent, great idea! Though if you have a border around the text you'd have to adjust the the block's padding so it wouldn't clip out of it. However it looks like you got around that by setting the first-line-indent to be 2em already! I'll try it out sometime, thanks for sharing :)

**Comment by pdcxs** [comment]
> I found another para `hanging-indent` if you want to indent first line in 1 em
> 
> ```
> set par(justify: false, leading: 1.5em, first-line-indent:2em )
> ```
> 
> you can do like this:
> 
> ```
> set par( hanging-indent:-1em )
> ```

This works anyway. But it will also affect the raw texts/code blocks.

**Comment by quanengineering** [comment]
This one also does the trick (thanks https://github.com/shuosc/SHU-Bachelor-Thesis-Typst/issues/12#issuecomment-1506872937)

```
show heading: it =>  {
    it
    par()[#text(size:0.5em)[#h(0.0em)]]
}
```

**Comment by tcztzy** [comment]
> This one also does the trick (thanks [shuosc/SHU-Bachelor-Thesis-Typst#12 (comment)](https://github.com/shuosc/SHU-Bachelor-Thesis-Typst/issues/12#issuecomment-1506872937))
> 
> ```
> show heading: it =>  {
>     it
>     par()[#text(size:0.5em)[#h(0.0em)]]
> }
> ```

so quick and so dirty, but I like it 👍

**Comment by teenjuna** [comment]
@laurmaedje sorry for bothering, but do you know if there are plans to resolve this?

**Comment by datleviet** [comment]
My temporary solution:

![image](https://github.com/typst/typst/assets/69759053/7c9f7369-eaab-48de-83ba-89200818d655)
![image](https://github.com/typst/typst/assets/69759053/52ec03f6-c487-4831-a32a-1a744cc96ecf)

**Comment by laurmaedje** [comment]
@teenjuna There are definitely plans to resolve it. Before we can do it, we have to have a stronger model of what is a paragraph and what is just text. That requires some deeper changes to how Typst works internally. The same changes are also relevant for HTML export. It will still take some time, but it is still on my radar.

**Comment by teenjuna** [comment]
@laurmaedje thanks for response! Good to know this is also related to HTML renderer, as it insures that this issue won't stale.

My 2 cents: I would expect Typst to have a separate Text element that just holds text. It would be used inside lists, equations, etc. Paragraph would then me just a container for top-level text on a page and can have styles that won't affect just Text.

In HTML, such Text element would map to <span>, which is **not** semantically incorrect according to MDN (as span element doesn't have special meaning).

**Comment by Andrew15-5** [comment]
> in Brazilian Portuguese documents we are used to apply indentation to every paragraph

Same goes with Russian.


> ```typ
> show heading: it =>  {
>     it
>     par()[#text(size:0.5em)[#h(0.0em)]]
> }
> ```

This can be optimized like so:

```typ
show heading: it => {
  it
  par(text(size: 0.5em, "")) // using [] instead of "" isn't the same
}
```

But this introduces additional vertical spacing, which isn't a desired result.

**Comment by Andrew15-5** [comment]
Surprisingly (to me), this works perfectly with no issues (maybe there are some):

```typ
#let indent = 1.25cm // Russian standard AFAIK
#set par(first-line-indent: indent)
#show heading: it => { it; h(indent) }
```

or

```typ
#let indent = 1.25cm
#set par(first-line-indent: indent)
#show heading: it => {
  it
  h(indent)
}
```

**Comment by Andrew15-5** [comment]
> ```typ
> #let annotation(body)={
> par(
>   [#h(3em)#body]
> )
> }
> ```

This is the safest and easiest approach for now. Here is a simplified version:

```typ
#let indent = 1.25cm
#set par(first-line-indent: indent)

#let indent_par(body) = par(h(indent) + body)
// or
#let indent_par(body) = par[#h(indent)#body]
```

Then just do:

```typ
= Heading
#indent_par[any text or #lorem(30)] // First paragraph

#lorem(30) // nth paragraph
...

// or
= Heading
#indent_par(lorem(30) // First paragraph

#lorem(30) // nth paragraph
...

**Comment by Andrew15-5** [comment]
> I found another para `hanging-indent` if you want to indent first line in 1 em
> 
> ```
> set par(justify: false, leading: 1.5em, first-line-indent:2em )
> ```
> 
> you can do like this:
> 
> ```
> set par( hanging-indent:-1em )
> ```

Maybe it worked in previous versions, but now (`0.7`) it doesn't (text goes out of the default margins):

![image](https://github.com/typst/typst/assets/37143421/2a7531b8-62fe-4594-a845-33d73fa5ff9a)

<details><summary>source code</summary>

```typ
#set par(first-line-indent: 2em)

= Heading
#lorem(30)

#lorem(30)

#lorem(30)

#set par(first-line-indent: 2em, hanging-indent: -1em)

= Heading
#lorem(30)

#lorem(30)

#lorem(30)
```

</details>

Maybe some additional settings are necessary.

**Comment by Andrew15-5** [comment]
One other way is to set a `show` rule, but you can't undo in v0.7:

```typ
#show "|": h(indent) // |
#show ">": h(indent) // >
#show "~": h(indent) // \~
```

You can choose almost any Unicode character, but I think that letters are easier to use (since it's the intended way of using the show rule). I really would like to use some fancy syntax for show rules, but you can't do that right now. And `regex("")` is therefore mostly useless.

**Comment by PuntitOwO** [comment]
Maybe add a note to the [Paragraph documentation](https://typst.app/docs/reference/layout/par/) that the following workarounds can be used to get the desired behavior? 

I suggest adding these three

```typst
// Workaround 1: Manually add horizontal space
#let indent = 3em // or whatever value you want
#let indent_par(body) = par[#h(indent)#body]
// Usage:
#indent_par[First paragraph]

// Workaround 2: Special character
#show "¬": h(3em) // Use whatever symbol and space you want
// Usage:
¬First paragraph

// Workaround 3: Automatically add empty paragraph after heading
#show heading: it => {
    it
    par(text(size:0.35em, h(0.0em)))
} // Only works for paragraphs directly after heading
```

**Comment by Andrew15-5** [comment]
And what did this "unfix" then?

**Comment by Enter-tainer** [comment]
IDK, i don't understand what `consecutive` means. And here seems to be some explaination https://github.com/typst/typst/issues/2350#issuecomment-1754422551

**Comment by Andrew15-5** [comment]
Maybe @PgBiel or @laurmaedje can explain. I, too, don't understand what's the point of `consecutive` right now.

**Comment by laurmaedje** [comment]
Consecutive is what indents only the second and following paragraphs. Just removing it won't have the intended effect though because suddenly everything would be indented: Headings, captions, single worlds in boxes, etc.

Try writing `A #box[B] C` in your fixed version. B will also be indented.

The fundamental problem we need to fix first is that text can't exist outside of a paragraph at the moment.

**Comment by Andrew15-5** [comment]
> The fundamental problem we need to fix first is that text can't exist outside of a paragraph at the moment.

Let's clear something here. Can text exist outside of a paragraph in Typst 0.9?

If I understand correctly, this statement must be true eventually:

> text can't exist outside of a paragraph

**Comment by laurmaedje** [comment]
In all Typst versions to date, text will always be layouted as part of a paragraph. In the future, we have to define some separation because not all text should be affected by paragraph indent.

**Comment by Andrew15-5** [comment]
> In the future, we have to define some separation

"Some separation" means a new element type or something? Or is this the main problem: to define "some separation" exactly?

**Comment by laurmaedje** [comment]
That's the main problem.

**Comment by Andrew15-5** [comment]
> In all Typst versions to date, text will always be layouted as part of a paragraph

So any text in any mode will be wrapped in a `par` element, correct? If I explicitly wrap text in a `par` element, will it be wrapped twice or the implicit `par` element will be substituted by the explicit `par` element?

In some documents, text can be used just as a single word inserted in a specific place/position, i.e., not as a text paragraph. Therefore, such example shouldn't be affected by indent, hence shouldn't implicitly be wrapped in `par` element. This also refers to `heading`, `box` element and probably much more.

This makes me think that the only 2 instances where there is a paragraph (that need to be indented) are these:

```typ
Just a simple and small paragraph in markup mode.

#par[Another paragraph that is wrapped in explicit `par` element.]
```

Wouldn't this fix the problem? Only apply implicit `par` to a text in markup or if the `par` used explicitly. In any other circumstances, `par` won't be used, and no indent will be applied. Sounds to me like a case closed, just need to fix some logic in Rust. What am I missing here?

Or in the current version of Typst, you can't physically "not use `par`" because it is hard-coded in every single element? If so, then `par` and its use must be re-thinked and re-done (re-written).

**Comment by Andrew15-5** [comment]
> ```typst
> Just a simple and small paragraph in markup mode.
> 
> #par[Another paragraph that is wrapped in explicit `par` element.]
> ```

One problem that I see here is inline insertion of something into the pure text paragraph, specifically at the very beginning of the paragraph. I think that if this **something** is not another element or does not contain another element (referring to `content`), then we make it a part of a pure text paragraph (mixed paragraph). Meaning that in such case, indent will be present. But if this **something** is another element or contain such, then indent won't be present.

Here are some examples (where `→` is Typst-created paragraph indentation):

```typ
→Just a simple and small paragraph in markup mode.
→Just a simple and #string small #integer paragraph #array-with-some-simple-type-values in markup mode.
→Just a simple and small #grid([`block` element splits 1 paragraph into 2])→paragraph in markup mode.
→#[content without elements] Just a simple and small paragraph in #content-value markup mode.
#box[explicit non-`par` element removes indent/implicit `par`]Just a simple and small paragraph in markup mode.
→#par[Just a simple and small paragraph in markup mode.]
→#par[This will probably create a separate paragraph and move text on the right to the 2nd implicit paragraph (like it is in Typst 0.9).]→Just a simple and small paragraph in markup mode.
→#{
  for index in range(5) {
    str(index)
  }
} Just a simple and small paragraph in markup mode.
#table([table is `block`, so text on the right will go to a new line, therefore a new paragraph.])→Just a simple and small paragraph in markup mode.
→Just a simple and #box(table([inline table])) small paragraph in markup mode.
```

The examples above just show how I see the logic work in practice if this issue (#311) is fixed, and I think all of them obey my <strike>simple</strike> rules stated above:
Add indent (make a `par` element) if:
1. pure text in root markup mode;
2. `par` is explicitly used to wrap "paragraph text";
3. text with inline (`box`) elements not at the beginning of a paragraph (not at the 1st character);
4. text with inline `content` (no `block`s inside) not at the beginning of a paragraph (not at the 1st character);
5. text with `block` (not inline) elements not at the beginning of a paragraph (not at the 1st character), **BUT** such elements will be put on a newline and all the text after these elements will be also put on another newline, therefore creating a new `par`;
6. text with simple type values at the beginning of a paragraph (at the 1st character).

I'm not sure about inline elements at the beginning of a paragraph, like this one:

```typ
#box[explicit non-`par` element removes indent/implicit `par`]Just a simple and small paragraph in markup mode.
```

Maybe this should automatically be converted to `par`. I can imagine an (inline) image/icon/custom bullet at the beginning of a paragraph, so the indent should be present in such cases.

**Comment by matthew-e-brown** [comment]
> I'm not sure about inline elements at the beginning of a paragraph, like this one:
>
> ```typst
> #box[explicit non-`par` element removes indent/implicit `par`]Just a simple and small paragraph in markup mode.
> ```

My intuition is that inline content at the start of a paragraph should **not** have any affect on indentation. After all, that's what it means to be inline: it's **inside** the paragraph&mdash;it just happens to be at the beginning. What if I wanted to start my paragraph off with a little icon or logo, for whatever reason?

If you wanted to remove implicit indentation before inline content, you would just use the exact same method that applied to any other paragraph.

```typst
#let noindent = h(0pt, weak: true)

#noindent Some paragraph completely on its own.

#noindent #box[A paragraph that] has inline content at its start.
```

**Comment by Andrew15-5** [comment]
> If you wanted to remove implicit indentation before inline content, you would just use the exact same method that applied to any other paragraph

Sometime it would be very handy, but in some situations simple `#h(-5em)` doesn't work. There should be a forceful way (function) to always be able to remove (any) indentation. Another example is enumerated list (can't remove indent without breaking the numeration).

**Comment by matthew-e-brown** [comment]
Maybe straying a bit into "general discussion" territory here, but something I come across quite frequently, that Andrew also touched on, is putting block-level things within the flow of a paragraph.

For example, my only real gripe with Typst so far has been managing spacing around lists or enums that are between paragraphs:

```typst
Some paragraph that uses a list to introduce:
- One topic;
- Two topics; or
- Maybe even more topics!
And then concludes with a few more words down here.
```

As far as _layout_ is concerned, these are three block-level elements: `par`, `list`, `par`. But, as far as _writing_ is concerned, it contains one topic and is one single paragraph. I would expect the second `par` to have no extra indentation, and have regular line-spacing between it and the `list` (at least when the list is `tight`).

I have so far been unable to get Typst to reliably do this without putting `v(0pt)` spaces below tight lists, to consume the paragraph's `above`. That has the problem that now I can't _end_ a paragraph with a `list`, since now it will eat up the spacing between what should be the next paragraph (not that that's all too proper, grammatically speaking, but I have been using Typst for not-necessarily-formal things like assignments). I think Typst should make it both possible and easy to achieve that sort of spacing, if I so desire.

HTML and CSS have this same problem, too: given a sequence of just `<p>`, `<ul>`, `<p>` elements, it's impossible to know if the second `<p>` is the continuation of the first "paragraph," or the start of a following one. That is, without putting a `no-indent` class on the second one, or something.

In Typst, however, whitespace is significant: an empty line **means** something. Perhaps Typst could make the distinction by whether or not a blank line appears between the last list item and the following paragraph:

```typst
This paragraph starts here...
- And goes into a list...
Before returning to a paragraph, without indentation. Then, it
- Goes back into a list...
- And ends here (however informal that may be).

Because there's an empty line here, _this_ is a *whole new paragraph,* instead of a continuation of the previous one.
```

That syntax seems a bit... loose, though. Just bringing up a concern of mine, and throwing an idea out there. 😁

**Comment by matthew-e-brown** [comment]
> Sometime it would be very handy, but in some situations simple `#h(-5em)` doesn't work. There should be a forceful way (function) to always be able to remove (any) indentation. Another example is enumerated list (can't remove indent without breaking the numeration).

`h` _should_ work for that with the `weak` parameter, I think...

> https://typst.app/docs/reference/layout/h/#parameters-weak
> 
> Weak spacing in markup also causes all adjacent markup spaces to be removed, regardless of the amount of spacing inserted. To force a space next to weak spacing, you can explicitly write `#" "` (for a normal space) or `~` (for a non-breaking space). The latter can be useful to create a construct that always attaches to the preceding word with one non-breaking space, independently of whether a markup space existed in front or not.

That should "remove (any) indentation," shouldn't it? Or does it not work on paragraph indentation? 🤔

**Comment by laurmaedje** [comment]
@matthew-e-brown Before the list, this is already special-cased. Not at the bottom though. I think back then I figured that it only makes sense to consider a tight list at the end of a paragraph and continuing after that would be strange. Maybe it was the wrong call.

**Comment by Andrew15-5** [comment]
Ok, maybe I went too far with this. Generally, using a negative indentation with `#h()` does give the desired result, but it should be rather an `#undo-indent` function, because I don't want to either create a `#let undo-indent = h(-indent-value)` shortcut myself or use a bare (and ugly) `#h(-indent-value)` all the time. Because I really do want Typst to be super readable (that's kinda the point when comparing with LaTeX). But I did face some problems. This example works pretty good:

<details><summary>source</summary>
<p>

```typ
#set par(first-line-indent: 1em)
#set list(body-indent: 1em)

First paragraph.

Second paragraph.

#h(-1em)Third paragraph.

- item

  #h(-1em)Paragraph

  #h(-1em)#box(table(columns: 2, [cell], [cell]))

  Also need to subtract a length of "#sym.bullet" somehow (I can _add_ by using
  ```typ #hide(sym.bullet)```, which isn't ideal either) to make a perfect 0-level
  root indentation (no indentation):

  #h(-2em)#box(table(columns: 2, [cell], [cell]))
- item
```
</p>
</details> 

<details><summary>output</summary>
<p>

![image](https://github.com/typst/typst/assets/37143421/f58b196f-1720-4fa8-87a5-7d58a8bf3cb3)
</p>
</details> 

But here is a bug (cell content is not indented by the same value/distance from all 4 sides inside the cell):

<details><summary>source</summary>
<p>

```typ
#set par(first-line-indent: 1em)
#set list(body-indent: 1em)

- item

  #h(-2em)#box(table(columns: 2, [cell], [cell]))
```

</p>
</details> 

<details><summary>output</summary>
<p>

![image](https://github.com/typst/typst/assets/37143421/a18ab9d5-7729-4902-adbc-ed6ae3990246)

</p>
</details> 

I have another example, but I currently can't show it. Maybe later.

**Comment by Andrew15-5** [comment]
Here it is. I now figured out that I didn't add an empty line after the list item, so the table wasn't able to dedent.

<details><summary>source</summary>
<p>

```typ
+ word
  #h(-1em)#box(table(
    columns: (2fr, 2fr, 1fr),
    [a], [b], [c],
    ..range(3 * 7).map(_ => [~]),
  ))
+ word
``` 

</p>
</details> 

<details><summary>output</summary>
<p>

![image](https://github.com/typst/typst/assets/37143421/6b084329-5044-4356-a267-9f69b82237a2)

</p>
</details>

With an empty line:

<details><summary>source</summary>
<p>

```typ
+ word

  #h(-1em)#box(table(
    columns: (2fr, 2fr, 1fr),
    [a], [b], [c],
    ..range(3 * 7).map(_ => [~]),
  ))
+ word
``` 

</p>
</details> 

<details><summary>output</summary>
<p>

![image](https://github.com/typst/typst/assets/37143421/e6ec0a4c-51a1-43c3-92ff-4e37739aff57)


</p>
</details>

**Comment by HPDell** [comment]
Hi @araruna, @matthew-e-brown, and others,

As a work-around, I suggest the following code to custom first-line indent via some conditions:

```typst
#show: body => {
  for (ie, elem) in body.children.enumerate() {
    if elem.func() == text {
      if ie > 0 and body.children.at(ie - 1).func() == parbreak {
        h(2em)
      }
      elem
    } else {
      elem
    }
  }
}
```

And the following content 

```typst
= Section

#lorem(40)
+ abc
+ def
+ ghi
#lorem(20)
- asd
- wds
- wasd

#lorem(20)

== Subsection
```

will result in

![image](https://github.com/typst/typst/assets/17842185/ca7054a2-2141-4e64-b093-bf335e37c84c)

But I think it would be better if there is a parameter that we can control whether the first line in the immediately folllowing **paragraph**, not the paragraph after a `parbreak`, is indented.

**Comment by PPakalns** [comment]
One more workaround that works for headings, lists, figures, tables etc.

```
let indent_hack = it => {
        show par: set block(spacing: 0em)
        it; v(-1.15em); ""
}

set par(justify: true, first-line-indent: 2em)
show heading: it => {
    indent_hack(it)
    v(0.3em)
}
show list: indent_hack
show enum: indent_hack
show table: indent_hack
show grid: indent_hack
show rect: indent_hack
show figure: indent_hack

# Add other display functions after which line should be with indent
```

**Comment by BrandonHenke** [comment]
Hey, friends. This has been an interesting issue, so I'll try to explain what's happening.

Paragraphs only contain "inline" elements. This includes
```
content.is::<SpaceElem>()
            || content.is::<TextElem>()
            || content.is::<HElem>()
            || content.is::<LinebreakElem>()
            || content.is::<SmartQuoteElem>()
            || content.to_packed::<EquationElem>()
                .is_some_and(|elem| !elem.block(styles))
            || content.is::<BoxElem>()
```
(Text, horizontal spacing, linebreaks, smart quotes, inline equations, and boxes).

This means, if any element other than one of those shows up, it isn't included in a paragraph. Suppose you have two paragraphs and you want to put a block equation in the first one. Before the block equation is put in, Typst simply has two paragraphs. After the equation is put in, however, typst splits the first paragraph into two: one before the equation and one after the equation. Now, there are essentially four blocks:
- par1(1/2)
- eq
- par1(2/2)
- par(2)
This also happens for lists, headers, anything that isn't one of the elements I've listed above.

Now, on top of this, a paragraph is only indented if it follows another paragraph. This is the `consecutive: bool` variable. If a paragraph follows another paragraph, `consecutive = true`. Therefore, using any item other than the ones I listed splits the paragraph in two, the second of which will not be indented because it doesn't technically follow another paragraph.

It's quite easy to add an argument `always-indent-first-line: bool` to the paragraph item, which is something I've done in my forked repo. However, this means that an indent will occur on everything that Typst considers a paragraph. For example,

```
Consider a manifold, $cal(M)$, and the ordinary differential equation,// This is its own paragraph, so it's indented.
$
  (d x) / (d t) = u (x(t))
$
where // This is its own paragraph, so it's indented.
$
  u: cal(M) -> T cal(M)
$
is a vector field on the manifold. // This is its own paragraph, so it's indented.
```

Even though all of this is one contiguous thought, the way Typst lays this out is odd to the everyday English writer. My proposal is to allow non-inline elements to be included in paragraphs. I'm working on modifying that in my fork.

There are several other issues that are related to this. Here is one:
https://github.com/typst/typst/issues/3206#issue-2083875484

**Comment by Fanisting** [comment]
> Hi @araruna, @matthew-e-brown, and others,
> 
> As a work-around, I suggest the following code to custom first-line indent via some conditions:
> 
> ```typst
> #show: body => {
>   for (ie, elem) in body.children.enumerate() {
>     if elem.func() == text {
>       if ie > 0 and body.children.at(ie - 1).func() == parbreak {
>         h(2em)
>       }
>       elem
>     } else {
>       elem
>     }
>   }
> }
> ```
> 
> And the following content
> 
> ```typst
> = Section
> 
> #lorem(40)
> + abc
> + def
> + ghi
> #lorem(20)
> - asd
> - wds
> - wasd
> 
> #lorem(20)
> 
> == Subsection
> ```
> 
> will result in
> 
> ![image](https://private-user-images.githubusercontent.com/17842185/320558520-ca7054a2-2141-4e64-b093-bf335e37c84c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3NDY1NjUsIm5iZiI6MTcxNDc0NjI2NSwicGF0aCI6Ii8xNzg0MjE4NS8zMjA1NTg1MjAtY2E3MDU0YTItMjE0MS00ZTY0LWIwOTMtYmYzMzVlMzdjODRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDE0MjQyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1ZTdlZDE5Mjc1N2FiMjAyOTI0MjgxMWQ4MWE4OTFjNmMyMzk2ZTYwNmY2YjM0NmNhYzZlMmJlOGQxYThhMjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JaQOBmoMEtIP-R5--ZOvJZ8qSGpIq7ltoxZMFlkpZDI)
> 
> But I think it would be better if there is a parameter that we can control whether the first line in the immediately folllowing **paragraph**, not the paragraph after a `parbreak`, is indented.
![tmp5292](https://github.com/typst/typst/assets/87118530/f049a6b9-4dfa-4236-9d1c-8cc44d8e737a)


This word around is super amazing but I would like to add that the indentation would not come into effect if you are using some styling options like bold. See the screenshot for a toy example. But simply, add `#h(2em)`  before the not-indented paragraph can do a good job. 

```
= Example heading

Some text here.

#h(2em) _Here_: some bold text at start.
```

**Comment by HPDell** [comment]
Hi @Fanisting,

The issue you mentioned occurs because the leading element is `strong` , not text. So the indent condition `elem.func() == text and ie > 0 and body.children.at(ie - 1).func() == parbreak` cannot be true in this case. To fix this issue, a better indent condition can be applied to cover this special case.

**Comment by Andrew15-5** [comment]
I started a new project which is a good reason to finally try and stop using `〜` as a manual intent after every heading, list, figure etc. and instead test workarounds that were posted here. Maybe I missed something, but the only good one was posted by @HPDell, because other workarounds mess up the layout with vertical or horizontal spacing.

I initially used it as is, and it was already very nice to have automatic indentation _everywhere_. But soon I faced my first problem: when inserting medium `[]` it is not flattened and therefore adds a nested structure ("subdocument"). But thankfully by adding 2 lines of code I was able to fix it using a recursive show rule function.

And soon enough, some places started misbehaving. In the end, one place where I can test several situations started to work as expected no matter what I did. But I still don't have anything besides paragraphs and headings, so my current solution will probably break again in the future. But for now this looks robust enough:

<details><summary>snippet</summary>

```typ
#let indent = 1em
#let styled = [#set text(red)].func()
#let space = [ ].func()
#let sequence = [].func()

#let turn-on-first-line-indentation(
  doc,
  last-is-heading: false, // space and parbreak are ignored
  indent-already-added: false,
) = {
  for (i, elem) in doc.children.enumerate() {
    let element = elem.func()
    if element == text {
      let previous-elem = doc.children.at(i - 1)
      if i == 0 or last-is-heading or previous-elem.func() == parbreak {
        if not indent-already-added {
          indent-already-added = true
          h(indent)
        }
      }
      elem
    } else if element == heading {
      indent-already-added = false
      last-is-heading = true
      elem
    } else if element == space {
      elem
    } else if element == parbreak {
      indent-already-added = false
      elem
    } else if element == sequence {
      turn-on-first-line-indentation(
        elem,
        last-is-heading: last-is-heading,
        indent-already-added: indent-already-added,
      )
    } else if element == styled {
      styled(turn-on-first-line-indentation(
        elem.child,
        last-is-heading: last-is-heading,
        indent-already-added: indent-already-added,
      ), elem.styles)
    } else {
      indent-already-added = false
      last-is-heading = false
      elem
    }
  }
}

#show: turn-on-first-line-indentation
```

</details>

P.S. There is little logic to it: I was just checking the output and trying to make a bunch of conditions, so that there will be just one indentation at a single place in a document (paragraph).

**Comment by HPDell** [comment]
Hi @Andrew15-5 ,

Thanks for your workaround which seems powerful and is able to cover more situations.

After posting my solution, I was thinking that is there any possibility to set up an array of conditions and check their results with functions like `any()` and `all()` to avoid too many if-else structure. So that users can add their own conditions to control the indent behaviour.

**Comment by Andrew15-5** [comment]
The outer if conditions are all for element type checking, which means that you probably can add only maybe 15 more, which is a lot, but not infinite (and most wouldn't be big anyway). To combine multiple elements under one condition, you can do `if element in (table, figure, raw) {}`. I tried making a switch case thingy, but it was far less readable.

**Comment by Andrew15-5** [comment]
I'm pretty sure in some standards/countries not having a first-line indentation is the correct way.

**Comment by hurzl** [comment]
I'm pretty sure I have never seen the indent missing on a new page.

**Comment by calebeby** [comment]
I found this package that someone made to workaround this problem: https://github.com/flaribbit/indenta. It seems to work pretty well.

**Comment by MikhailKravets** [comment]
This simple trick `#{ "" }` works for me:

```typst
#set par(first-line-indent: 1.5em, justify: true, leading: 1.5em)

== First one

// This make the first line of the first paragraph to be indented
#{ "" }

#lorem(200)

#lorem(200)
```

**Comment by Andrew15-5** [comment]
This and many other "tricks" and "workarounds" create an additional vertical spacing/indentation that can be unacceptable for some people (if you have strict guidelines, or you're a perfectionist). And if you're still going to use it, then you can also use `#""` instead.

<details><summary>More about adding vertical spacing (near headings) rather than horizontal indentation, but there is still some overlap.</summary>

That's why I now only use root show rule for this one. You have a full control over what is being inserted and where (https://github.com/typst/typst/issues/311#issuecomment-2104447655). The only problem I found with a few of my designs, is that I can't trace the state of the variables after a nested `sequence`. I _think_ that it's fixable by returning state/variables in addition to the element, but I haven't tried it. This means that the current design doesn't work with nested blocks of content, a.k.a. a file with `#include` statement(s). Which in turn forced me to use a flat file structure, otherwise I will have incorrect spacing near headings where `#include`s connect.

</details>

**Comment by iiztp** [comment]
#show heading: it => [
  #it.numbering
  #it.body
  #{""}
]

> This simple trick `#{ "" }` works for me:
> 
> ```typst
> #set par(first-line-indent: 1.5em, justify: true, leading: 1.5em)
> 
> == First one
> 
> // This make the first line of the first paragraph to be indented
> #{ "" }
> 
> #lorem(200)
> 
> #lorem(200)
> ```

Following your solution, this fixes works : 
```
#show heading: it => [
  #it
  #{""}
]
```

**Comment by lucas-bublitz** [comment]
For the perfectionists, add a negative vertical spacing after creating a new paragraph. I do not like this because it reminds me of the LaTeX quick fixes (_gambiarras_ in my language), which I am trying to avoid using Typst.
```
show heading: it => {
    set text(12pt)
    it
    ""
    v(-1em)
}
```

**Comment by jsx97** [comment]
Personally I prefer `h(1em)` for paragraphs and `pad(left: 1em)` for lists.

```
#set par(first-line-indent: 1em, justify: true, leading: 1em)

= lorem

#lorem(20)

#lorem(20)

#pad(left: 1em, [
  + #lorem(20)
  + #lorem(20)
])

#h(1em)#lorem(20)

#lorem(20)
```

**Comment by MirionQs** [comment]
just add `show heading: el => el + ""`

**Comment by SandaruKasa** [comment]
Let me add my two cents of a workaround as well
```typst
#show heading: it => [ #it #v(-par.spacing) #{ "" } ]
#show figure: it => [ #it #v(-par.spacing) #{ "" } ]
#show terms: it => [ #it #v(-par.spacing) #{ "" } ]
#show list: it => [ #it #v(-par.spacing) #{ "" } ]
```

**Comment by PgBiel** [comment]
> Let me add my two cents of a workaround as well

For future reference, I've posted a similar solution in the forum as well, with examples: https://forum.typst.app/t/how-to-indent-every-paragraph-excluding-headings/2137/4

It boils down to:

```typ
// Add more elements below as necessary, e.g. list
#show selector.or(heading, table, grid, figure): it => {
    it
    ""
    context v(-par.spacing - measure("").height)
}
```

It's worth noting that you can unify the show rules by using `selector.or` as above. Also, for consistency with the original heading spacing, you should exclude the spacing for `""`.

**Comment by Andrew15-5** [comment]
Are there any known or potential issues with this particular method? Perhaps if some set of settings is changed, meaning the state is nothing like default settings? I'm thinking of a fuzzy testing of sorts. All because it's so simple yet so powerful/robust (unlike all others), at least on the paper. I'll put it to the test with my next documents.

**Comment by fruzitent** [comment]
> Let me add my two cents of a workaround as well
> 
> ```typst
> #show heading: it => [ #it #v(-par.spacing) #{ "" } ]
> #show figure: it => [ #it #v(-par.spacing) #{ "" } ]
> #show terms: it => [ #it #v(-par.spacing) #{ "" } ]
> #show list: it => [ #it #v(-par.spacing) #{ "" } ]
> ```

I had to wrap the figure in a `#block[]` (`#box[]` prevents breakable tables) because large image can create an extra blank page.

![image_2024-12-29_19-51-56](https://github.com/user-attachments/assets/bd1ff5da-cde8-4920-a8e7-705ff24a5c7e)
![image_2024-12-29_19-51-56 (2)](https://github.com/user-attachments/assets/c6b4af7a-aaf4-4c3b-b337-38262c5a8d5c)

**Comment by Andrew15-5** [comment]
I found that it doesn't work after `#pagebreak()`:

```typ
#let indent = 1.25cm
#set list(indent: indent)
#set par(first-line-indent: indent)
#show heading: set align(center)
#let complex-title-page(title: none) = align(center + horizon)[#title]

#show selector.or(heading, list, enum, figure, page): it => {
  it
  ""
  context v(-par.spacing - measure("").height)
}

#complex-title-page(title: "My title")
#pagebreak()

#lorem(20)

#lorem(20)

- #lorem(5);
- #lorem(5);
- #lorem(5).

#lorem(20)

== Heading
#lorem(20)
#figure(rect(), caption: [])
#lorem(20)
```

![image](https://github.com/user-attachments/assets/af283dcd-f664-4ae0-b2c8-dc96d0188da9)

**Comment by PgBiel** [comment]
Correct; in general, it requires something to be before the paragraph.

It isn't very easy to solve that without a native implementation. But I think it's OK to manually handle those specific cases for the time being.

**Comment by Andrew15-5** [comment]
> But I think it's OK to manually handle those specific cases for the time being.

`#show "〜": h(indent)` is like:

<img height="110px" src="https://github.com/user-attachments/assets/a6c45284-8c42-45b5-8052-eb44eeda8612" />

**Comment by danilasar** [comment]
> I started a new project which is a good reason to finally try and stop using `〜` as a manual intent after every heading, list, figure etc. and instead test workarounds that were posted here. Maybe I missed something, but the only good one was posted by @HPDell, because other workarounds mess up the layout with vertical or horizontal spacing.
> 
> I initially used it as is, and it was already very nice to have automatic indentation _everywhere_. But soon I faced my first problem: when inserting medium `[]` it is not flattened and therefore adds a nested structure ("subdocument"). But thankfully by adding 2 lines of code I was able to fix it using a recursive show rule function.
> 
> And soon enough, some places started misbehaving. In the end, one place where I can test several situations started to work as expected no matter what I did. But I still don't have anything besides paragraphs and headings, so my current solution will probably break again in the future. But for now this looks robust enough:
> 
> snippet
> ```typst
> #let indent = 1em
> #let styled = [#set text(red)].func()
> #let space = [ ].func()
> #let sequence = [].func()
> 
> #let turn-on-first-line-indentation(
>   doc,
>   last-is-heading: false, // space and parbreak are ignored
>   indent-already-added: false,
> ) = {
>   for (i, elem) in doc.children.enumerate() {
>     let element = elem.func()
>     if element == text {
>       let previous-elem = doc.children.at(i - 1)
>       if i == 0 or last-is-heading or previous-elem.func() == parbreak {
>         if not indent-already-added {
>           indent-already-added = true
>           h(indent)
>         }
>       }
>       elem
>     } else if element == heading {
>       indent-already-added = false
>       last-is-heading = true
>       elem
>     } else if element == space {
>       elem
>     } else if element == parbreak {
>       indent-already-added = false
>       elem
>     } else if element == sequence {
>       turn-on-first-line-indentation(
>         elem,
>         last-is-heading: last-is-heading,
>         indent-already-added: indent-already-added,
>       )
>     } else if element == styled {
>       styled(turn-on-first-line-indentation(
>         elem.child,
>         last-is-heading: last-is-heading,
>         indent-already-added: indent-already-added,
>       ), elem.styles)
>     } else {
>       indent-already-added = false
>       last-is-heading = false
>       elem
>     }
>   }
> }
> 
> #show: turn-on-first-line-indentation
> ```
> 
> P.S. There is little logic to it: I was just checking the output and trying to make a bunch of conditions, so that there will be just one indentation at a single place in a document (paragraph).

Are you sure this code works? It looks strange when element type is sequence or styled. My friend used this solution with @preview/codelst:2.0.2 and that led to an error `text does not have field "children"` at the cycle declaration line (`for (i, elem) in doc.children.enumerate()`). It looks like when something element was styled, its child was text and recursion call of turn-on-first-line-indentation led to this error.

**Comment by Andrew15-5** [comment]
That workaround is very cursed and highly depends on your document structure. It uses a brute-force like approach of checking every scenario in a document. Just don't use it, or create your own from scratch for your document. Or good luck fixing my mess! ;)

For now, I switched to the superior in most ways solution: https://github.com/typst/typst/issues/311#issuecomment-2564502920. Although I do still need my regular solution in some rare cases: https://github.com/typst/typst/issues/311#issuecomment-2565881078. Now my setting preamble _and_ document content are both much more clean: the show rule is super short and almost never need to prefix any paragraph.

**Comment by danilasar** [comment]
Thanks for reply. I think it would be nice to have more than one solution so I'll try to do something with your code anyway. https://github.com/typst/typst/issues/311#issuecomment-2565881078 looks sad, but I liked  https://github.com/typst/typst/issues/311#issuecomment-2564502920

**Comment by jtchen2k** [comment]
Returning to this in 2025. Just curious, if the new layout engine released in 0.12 simplifies implementing this feature? I assume many non-English users are anticipating this(The hack of appending a fake paragraph under the headings could lead to orphan headings).

**Comment by laurmaedje** [comment]
There is progress on the feature. I can't make promises, but we are optimistic that this should be fixed this soon.

**Comment by Enivex** [comment]
Closed by https://github.com/typst/typst/pull/5768
